rfcs/convergence/simplify-prop-merging.md

```diff
@@ -0,0 +1,358 @@
+const state = mergeProps(
+  { button: { handleClick /* a require handler by our component */ } },
+   resolveShorthandProps(props),
+);
+// Results in => { button: { customHandleClick /* 💣 handlers are not merged */ } }
+```
+
+- When merging different types, simply replaces most recent usage. Although not a problem with slots because of `resolveShorthandProps`
+
+## Problem statement
+
+### Broken `as` prop
+
+Let's take a simple example with a `Loader` component slot inside a `Button` component.
+
+```tsx
+// `Button` state hook
+export const useButton = () => {
+  const state = mergeProps({
+    // ...
+    // Loader component used as shorthand slot
+    loader: { as: Loader },
+  });
+};
+
+// `Button` render function
+export function renderButton(state) {
+  const { slots, slotProps } = getSlots(state);
+  return (
+    <button>
+      {/* ... */}
+      {/* 👇 "slots.loader" will render Loader component */}
+      <slots.loader {...slotProps.loader} />
+    </button>
+  );
+}
+```
+
+```tsx
+// During usage, want to use `Loader` as `div`
+<Button loader={{ as: 'div' }} />
+```
+
+The result is rather worrying:
+
+```tsx
+// Expected result
+<Button>
+  <Loader as="div" />
+</Button>
+
+// Result
+<Button>
+  {/* 👇 this is just a plain div, Loader is not rendered 🤨 */}
+  <div />
+</Button>
+```
+
+In the above example `as` is intended to represent the `Loader` prop, not to defined the component to be rendered. This is a result of special handling of the `as` prop which will always be omitted from the slot props and used as the component to be rendered.
+
+### Complex typings and unsafe internal implementation
+
+[microsoft/fluentui#17508](https://github.com/microsoft/fluentui/pull/17508/files) includes type fixes for `mergeProps` which provides extra type safety for to validate shorthand slots.
+
+However, the result still leads to complicated component state types to benefit from the stricter typing through generics.
+
+### Native attribute filtering needs to happen by default
+
+For slots that render native elements, it's difficult to provide type safety in order to avoid unsafe DOM attribute spreads as shown in the below example:
+
+```tsx
+// ComponentState
+const state = {
+  buttonSlot: { as: 'button', type: 'submit' },
+};
+
+function App() {
+  // Typescript cannot warn that this is potentially evil
+  return <Component buttonSlot={{ as: 'div' }} />;
+}
+```
+
+```html
+<!-- Thankfully still works, but props must be filtered on every render -->
+<div></div>
+```
+
+The above example still works thanks to the the native attribute filtering that is done as a pat of `getSlots`. However, these filters always need to be run when resolving shorthand slots. During each render, every prop of a component is iterated and re-filtered.
+
+The current filtering also needs to iterate all the allowed props during as a nested loop, although this could be optimized quite easily.
```

Comment:
```suggestion
The current filtering also needs to iterate all the allowed props during a nested loop, although this could be optimized quite easily.
```
Is the intention here to reflect that filtering is done in a nested loop?